### DISCUSSION TITLE
Global using namespace

### DISCUSSION THREAD

**Discussion by rick-de-water** [dotnet/csharplang#5292] [Language Ideas]
I was hoping that global using would finally provide a proper way to create aliases for generic types (like typedef/using in C++). Unfortunately there are multiple issues.

1. I cannot add them to a namespace. I can force these aliases into the global namespace, but not into their own namespace. This is becoming especially inconvenient with new generic features such as the generic math.
2. The aliases are only available in the current project. This means that every project will have to add the global using statements on its own.

What I can do now:
```csharp
// Only available in this project
global using Vector2f = Lina.Vector2<float>;
global using Vector2d = Lina.Vector2<double>;
global using Vector2i = Lina.Vector2<int>;
global using Vector2u = Lina.Vector2<uint>;
```

What I want to be able to do:
```csharp
namespace Lina
{
    // Available in all projects
    global using Vector2f = Vector2<float>;
    global using Vector2d = Vector2<double>;
    global using Vector2i = Vector2<int>;
    global using Vector2u = Vector2<uint>;
}
```

Are there any plans to support proper aliasing of types?

**Comment by Mrxx99** [comment]
You can add the usings to a Directory.Build.props file so it is applied to every project in the folder of that file and below

**Reply by rick-de-water** [reply]
That would require every consumer of a library to manually define the aliases, which is not ideal at all

**Reply by jnm2** [reply]
The library author can include a .props file and/or a .targets file in their NuGet package: https://docs.microsoft.com/en-us/nuget/create-packages/creating-a-package#include-msbuild-props-and-targets-in-a-package

**Reply by msedi** [reply]
I already discussed this a years ago here: https://stackoverflow.com/questions/61813506/implicit-assembly-wide-using-directive?noredirect=1#comment109334969_61813506, I needed this for extension methods, since it is quite a nuisance to always add `using` for extension methods. Especially when you don't know that they are available. I would have needed the `using` addition automatically to make it part of the method list whenever the assembly is added as a reference or project reference.

@jnm2 suggestions could be a solution, but only if the package is distributed via nuget. It would be very helpful if this could be available when referencing an assembly with Reference or Project Reference too.

**Reply by CyrusNajmabadi** [reply]
>  I needed this for extension methods, since it is quite a nuisance to always add using for extension methods.

If you want all downstream consumers of your package to pick up your extension methods (Without adding a `using`) then this has always been possible.  You simply put your extensions in the global namespace.  Then users will exactly get this behavior.

Yes, this pollutes the global namespace with these extensions... but that's exactly what you're asking for here.  So it's no worse than any solution that involved exporting some sort of 'global using' that people pick up when they reference your package.

**Reply by msedi** [reply]
Is this also related to: #4029?

**Reply by msedi** [reply]
@CyrusNajmabadi: Thanks - thats a very good hint, I have never thought about this.

**Reply by rick-de-water** [reply]
I would still like there to be a namespace around the alias as well.
